### PR TITLE
add catalog option for cpd4 service install + fix custom resource plural + add dods service install

### PR DESCRIPTION
**Pull request by frederic-pesquet-work**
This PR:
- adds an option '--catalog-source _catalog_' to the command 'cpd4 service install', allowing to choose the catalog to install the service from (by default it is 'ibm-operator-catalog' as before)
- fix computation of custom resource plural: if custom resource name ends with 's', don't add another one in plural. This is the case for 'dods', whose plural is 'dods'
- fix wait for custom resource creation during service install: was crashing with 'Error: 'dict' object has no attribute 'metadata'' when resource was already created
- add 'ibm-dods' to the list of services you can install

This is a first step toward the installation of dev versions of services. You can:
- install a dev catalog (without dg for the moment, the integration will be in another command, to come)
- install the service from this catalog using the new --catalog-name option

### CODE REVIEW

File: dg/deps/config/cpd-subscriptions.json [hunk 1]
```diff
@@ -112,7 +112,7 @@
 			"name": "ibm-cpd-datastage-operator"
 		}
 	},
-	"ibm-cpd-dods": {
+	"ibm-cpd-dods-operator": {
```
Comment (jensmueller-com) [inline dg/deps/config/cpd-subscriptions.json:115]: Please revert this change as this file is machine-generated from the [documentation](https://www.ibm.com/docs/en/cloud-paks/cp-data/4.0?topic=tasks-creating-operator-subscriptions), which uses `ibm-cpd-dods` as the subscription name.

File: dg/lib/cloud_pak_for_data/cpd_4_0_0/cpd_service_manager.py [hunk 2]
```diff
@@ -164,5 +164,6 @@ def _initialize_subscriptions_dict_if_required(self):
                         spec=SubscriptionMetadataSpec(
                             channel=value["spec"]["channel"],
                             name=value["spec"]["name"],
+                            source=value["spec"].get("source", None)
```
Comment (jensmueller-com) [inline dg/lib/cloud_pak_for_data/cpd_4_0_0/cpd_service_manager.py:167]:
```suggestion
                            source=value["spec"].get("source")
```

File: dg/lib/cloud_pak_for_data/cpd_4_0_0/types/custom_resource_metadata.py [hunk 3]
```diff
@@ -137,8 +137,9 @@ def get_kind_metadata(self) -> KindMetadata:
         KindMetadata
             kind metadata
         """
+        plural = self._kind.lower() + ('s' if not self._kind.lower().endswith('s') else '')
```
Comment (jensmueller-com) [inline dg/lib/cloud_pak_for_data/cpd_4_0_0/types/custom_resource_metadata.py:140]: Add a newline after the docstring (see [coding guidelines](https://github.com/IBM/data-gate-cli/blob/master/docs/coding_guidelines.md#comments))

File: dg/lib/cloud_pak_for_data/cpd_4_0_0/cpd_manager.py [hunk 4]
```diff
@@ -772,7 +784,7 @@ def _create_operator_subscription(
 
         if not self._openshift_manager.subscription_exists(project, subscription_metadata.name):
             if len(created_dependent_subscriptions) == 0:
-                logging.info(f"Creating subscription '{subscription_metadata.name}' for operator '{operator_name}'")
+                logging.info(f"Creating subscription '{subscription_metadata.name}' for operator '{operator_name}' from catalog '{catalog_name}'")
```
Comment (jensmueller-com) [inline dg/lib/cloud_pak_for_data/cpd_4_0_0/cpd_manager.py:787]: - Remove the flake8(E501) warning by splitting the overlong string.
- In case you use Visual Studio Code, I would also suggest to format changed files with Black, which is configured in `.vscode/settings.json` (Command Palette → "Format Document" or enable `editor.formatOnSave` in settings for automatic formatting).

**Review by jensmueller-com — CHANGES_REQUESTED** [review]
Thanks a lot for this contribution, Frédéric!

File: .gitignore [hunk 5]
```diff
@@ -5,3 +5,5 @@ __pycache__/
 build/
 data_gate_cli.egg-info/
 dist/
+.eggs/
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply .gitignore:8]: Agree. Not sure how I did get there, but does not seem generic. I'll just remove them

File: dg/commands/cpd4/service/install.py [hunk 6]
```diff
@@ -70,6 +70,9 @@
         case_sensitive=False,
     ),
 )
+@click.option(
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply dg/commands/cpd4/service/install.py:73]: - agree with catalog source, changed
- sorting done
- long line split done

File: dg/commands/cpd4/service/install.py [hunk 7]
```diff
@@ -85,6 +88,7 @@ def install(
     service_name: str,
     storage_class: Optional[str],
     storage_vendor: Optional[str],
+    catalog_name: Optional[str],
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply dg/commands/cpd4/service/install.py:91]: done

File: dg/deps/config/cpd-custom-resources.json [hunk 8]
```diff
@@ -64,5 +64,21 @@
 		"status_key_name": "dmcAddonStatus",
 		"storage_option_required": false,
 		"version": "v1"
+	},
+	"ibm-dods": {
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply dg/deps/config/cpd-custom-resources.json:68]: done

File: dg/deps/config/cpd-custom-resources.json [hunk 9]
```diff
@@ -64,5 +64,21 @@
...
+		"group": "dods.cpd.ibm.com",
+		"description": "Decision Optimization",
+		"kind": "DODS",
+		"licenses": [
+			"Enterprise",
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply dg/deps/config/cpd-custom-resources.json:73]: WatsonStudioPremium was only in 4.0.0 and then deprecated by OMs. We don't want to support it, it is legacy

File: dg/deps/config/cpd-custom-resources.json [hunk 10]
```diff
@@ -64,5 +64,21 @@
...
+			"Standard"
+		],
+		"name": "dods-cr",
+		"operator_name": "ibm-cpd-dods-operator",
+		"spec": {			
+		},
```
[the remark this answers is not in the document]
Comment (frederic-pesquet-work) [reply dg/deps/config/cpd-custom-resources.json:79]: This is not required, and including it would defeat our short term purpose here, i.e. installing the latest version from a catalog.

File: dg/deps/config/cpd-custom-resources.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (jensmueller-com) [reply dg/deps/config/cpd-custom-resources.json:73]: In this case, it would make sense to open a ticket for the ID team to update the documentation.

File: dg/deps/config/cpd-custom-resources.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (jensmueller-com) [reply dg/deps/config/cpd-custom-resources.json:79]: That's good to know. I wonder why the documentation for some services includes this key.

[the remark this answers is not in the document]
Comment (jensmueller-com) [reply dg/deps/config/cpd-custom-resources.json:79]: Could you please remove the `version` key from the specifications for `datagate` and `dmc`? Also, please format the document. Thanks!

**Comment by frederic-pesquet-work** [comment]
@jensmueller-com I fixed all according to your (good) remarks. Can you review again please?

**Review by jensmueller-com — APPROVED** [review]
All good, thanks a lot!
